File: java/06_BatchFeedback.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17+
//DEPS com.fasterxml.jackson.core:jackson-databind:2.18.2

// Recipe 06: attach feedback to many Calls in one request.
//
// Demonstrates the bulk variant of /feedback/create:
//   POST /feedback/batch/create  -> N feedback items in one round trip
//
// Two wire-level points worth knowing:
//
//   - The path is /feedback/batch/create, not the more guessable
//     /feedback/create-batch or /feedback/createBatch.
//   - The body wraps a parallel-indexed array under `batch`:
//       {"batch": [<FeedbackCreateReq>, <FeedbackCreateReq>, ...]}
//     Each item carries its own project_id, weave_ref, feedback_type,
//     and payload — exactly the shape /feedback/create takes. The
//     response mirrors the input with {"res": [<FeedbackCreateRes>, ...]},
//     indices aligned to the input batch.
//
// This recipe creates three Calls and attaches two feedback items per
// Call in a single batch request: a wandb.note.1 (UI-visible in the
// trace table) and a custom scorer-style feedback. One round trip ships
// 6 items; the per-item endpoint would need 6. Mirrors recipe 05's
// note + scorer split, but bulk.
//
// Run:
//   jbang java/06_BatchFeedback.java

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

class BatchFeedback {
    static final ObjectMapper MAPPER = new ObjectMapper();
    static final HttpClient HTTP = HttpClient.newHttpClient();
    static final String BASE_URL = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai");
    static final String NOTE_TYPE = "wandb.note.1";
    static final String SCORER_TYPE = "recipe-06-scorer-correctness";
    static String projectId;
    static String authHeader;
    static Map<String, Object> baseAttributes;

    static String getenv(String key, String def) {
        String v = System.getenv(key);
        return (v == null || v.isEmpty()) ? def : v;
    }

    static void fail(String msg) {
        System.err.println("FAIL: " + msg);
        System.exit(1);
    }

    // Centralizes auth + JSON (de)serialization. Any non-2xx response is fatal.
    static JsonNode postJson(String path, Object body) throws Exception {
        String json = MAPPER.writeValueAsString(body);
        HttpRequest req = HttpRequest.newBuilder(URI.create(BASE_URL + path))
                .header("Authorization", authHeader)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
        HttpResponse<String> res = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2)
            fail("HTTP " + res.statusCode() + " for " + path + ": " + res.body());
        String b = res.body();
        return (b == null || b.isEmpty()) ? MAPPER.createObjectNode() : MAPPER.readTree(b);
    }

    static String startCall(String opName, Map<String, Object> inputs) throws Exception {
        JsonNode started = postJson("/call/start", Map.of(
                "start", Map.of(
                        "project_id", projectId,
                        "op_name", opName,
                        "started_at", Instant.now().toString(),
                        "attributes", baseAttributes,
                        "inputs", inputs)));
        return started.get("id").asText();
    }

    static void endCall(String callId, Map<String, Object> output) throws Exception {
        postJson("/call/end", Map.of(
                "end", Map.of(
                        "project_id", projectId,
                        "id", callId,
                        "ended_at", Instant.now().toString(),
                        "summary", Map.of(),
                        "output", output)));
    }

    public static void main(String[] args) throws Exception {
        var missing = new ArrayList<String>();
        for (String k : List.of("WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"))
            if (getenv(k, "").isEmpty()) missing.add(k);
        if (!missing.isEmpty())
            fail("Missing required env vars: " + String.join(", ", missing) + ". See ../README.md#setup.");

        projectId = System.getenv("WANDB_ENTITY") + "/" + System.getenv("WANDB_PROJECT");
        authHeader = "Basic " + Base64.getEncoder().encodeToString(
                ("api:" + System.getenv("WANDB_API_KEY")).getBytes(StandardCharsets.UTF_8));
        baseAttributes = new LinkedHashMap<>();
        baseAttributes.put("cookbook.language", "java");
        baseAttributes.put("cookbook.recipe", "06_batch_feedback");
        baseAttributes.put("cookbook.environment", getenv("COOKBOOK_ENVIRONMENT", "dev"));

        // Create three Calls — same shape as recipe 01, just repeated.
        String[][] questions = {
                {"What is the capital of France?", "Paris"},
                {"What is the capital of Spain?", "Madrid"},
                {"What is the capital of Italy?", "Rome"},
        };
        record Call(String id, String ref, String answer) {}
        var calls = new ArrayList<Call>();
        for (int i = 0; i < questions.length; i++) {
            String callId = startCall("recipe-06-call-" + (i + 1), Map.of("question", questions[i][0]));
            endCall(callId, Map.of("answer", questions[i][1]));
            calls.add(new Call(callId, "weave:///" + projectId + "/call/" + callId, questions[i][1]));
            System.out.println("Call " + (i + 1) + ": id=" + callId);
        }

        // Build the batch — note + scorer feedback per Call (6 items total).
        var batch = new ArrayList<Map<String, Object>>();
        for (Call c : calls) {
            batch.add(Map.of(
                    "project_id", projectId,
                    "weave_ref", c.ref(),
                    "feedback_type", NOTE_TYPE,
                    "payload", Map.of("note", "Reviewed — answer: '" + c.answer() + "'")));
            batch.add(Map.of(
                    "project_id", projectId,
                    "weave_ref", c.ref(),
                    "feedback_type", SCORER_TYPE,
                    "payload", Map.of("output", Map.of("score", 1.0, "reason", "Answer '" + c.answer() + "' matches expected"))));
        }

        // Single round trip for all six items.
        JsonNode resp = postJson("/feedback/batch/create", Map.of("batch", batch));
        JsonNode results = resp.path("res");
        if (results.size() != batch.size())
            fail("batch size mismatch: sent " + batch.size() + " got " + results.size());
        for (int i = 0; i < batch.size(); i++)
            System.out.println("Batch->Feedback: type=" + batch.get(i).get("feedback_type")
                    + " feedback_id=" + results.get(i).path("id").asText());

        // --- verification ---
        // For each Call, query feedback by weave_ref and assert both the note and
        // the scorer feedback landed with the expected payload. Brief retry
        // tolerates eventual consistency in the read path.
        List<String> expectedTypes = List.of(NOTE_TYPE, SCORER_TYPE);
        for (Call c : calls) {
            JsonNode expectedNote = MAPPER.valueToTree(Map.of("note", "Reviewed — answer: '" + c.answer() + "'"));
            JsonNode expectedScorer = MAPPER.valueToTree(Map.of("output", Map.of("score", 1.0, "reason", "Answer '" + c.answer() + "' matches expected")));
            var byType = new LinkedHashMap<String, JsonNode>();
            for (int i = 0; i < 5; i++) {
                JsonNode res = postJson("/feedback/query", Map.of(
                        "project_id", projectId,
                        "query", Map.of("$expr", Map.of("$eq", List.of(
                                Map.of("$getField", "weave_ref"),
                                Map.of("$literal", c.ref()))))));
                byType.clear();
                for (JsonNode row : res.path("result")) {
                    String t = row.path("feedback_type").asText(null);
                    if (t != null && expectedTypes.contains(t)) byType.put(t, row);
                }
                if (byType.keySet().containsAll(expectedTypes)) break;
                Thread.sleep(1000);
            }
            if (!byType.keySet().containsAll(expectedTypes))
                fail("feedback for " + c.ref() + " not all visible after 5 reads (got " + byType.keySet() + ")");
            if (!expectedNote.equals(byType.get(NOTE_TYPE).get("payload")))
                fail("note payload for " + c.id() + ": " + byType.get(NOTE_TYPE).get("payload"));
            if (!expectedScorer.equals(byType.get(SCORER_TYPE).get("payload")))
                fail("scorer payload for " + c.id() + ": " + byType.get(SCORER_TYPE).get("payload"));
            for (JsonNode row : byType.values())
                if (!c.ref().equals(row.path("weave_ref").asText()))
                    fail("weave_ref drift: " + row.get("weave_ref"));
        }
        System.out.println("Verified: " + batch.size() + " batched feedback items across " + calls.size() + " Calls (note + scorer each)");
    }
}

File: golang/06_batch_feedback.go
// Recipe 06: attach feedback to many Calls in one request.
//
// Demonstrates the bulk variant of /feedback/create:
//   POST /feedback/batch/create  -> N feedback items in one round trip
//
// Two wire-level points worth knowing:
//
//   - The path is /feedback/batch/create, not the more guessable
//     /feedback/create-batch or /feedback/createBatch.
//   - The body wraps a parallel-indexed array under `batch`:
//       {"batch": [<FeedbackCreateReq>, <FeedbackCreateReq>, ...]}
//     Each item carries its own project_id, weave_ref, feedback_type,
//     and payload — exactly the shape /feedback/create takes. The
//     response mirrors the input with {"res": [<FeedbackCreateRes>, ...]},
//     indices aligned to the input batch.
//
// This recipe creates three Calls and attaches two feedback items per
// Call in a single batch request: a wandb.note.1 (UI-visible in the
// trace table) and a custom scorer-style feedback. One round trip ships
// 6 items; the per-item endpoint would need 6. Mirrors recipe 05's
// note + scorer split, but bulk.
//
// Run:
//   go run golang/06_batch_feedback.go

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"reflect"
	"strings"
	"time"
)

const (
	noteType   = "wandb.note.1"
	scorerType = "recipe-06-scorer-correctness"
)

var (
	baseURL   = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai")
	projectID = os.Getenv("WANDB_ENTITY") + "/" + os.Getenv("WANDB_PROJECT")
	apiKey    = os.Getenv("WANDB_API_KEY")
	client    = &http.Client{}

	baseAttributes = map[string]any{
		"cookbook.language":    "golang",
		"cookbook.recipe":      "06_batch_feedback",
		"cookbook.environment": getenv("COOKBOOK_ENVIRONMENT", "dev"),
	}
)

func getenv(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func fatal(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}

// postJSON centralizes auth + JSON (de)serialization. Any non-2xx is fatal.
func postJSON(path string, body map[string]any) map[string]any {
	payload, err := json.Marshal(body)
	if err != nil {
		fatal("encode %s: %v", path, err)
	}
	req, err := http.NewRequest(http.MethodPost, baseURL+path, bytes.NewReader(payload))
	if err != nil {
		fatal("request %s: %v", path, err)
	}
	req.SetBasicAuth("api", apiKey)
	req.Header.Set("Content-Type", "application/json")
	res, err := client.Do(req)
	if err != nil {
		fatal("post %s: %v", path, err)
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(res.Body)
	if res.StatusCode/100 != 2 {
		fatal("HTTP %d for %s: %s", res.StatusCode, path, raw)
	}
	if len(raw) == 0 {
		return map[string]any{}
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		fatal("decode %s: %v", path, err)
	}
	return out
}

func startCall(opName string, inputs map[string]any) string {
	started := postJSON("/call/start", map[string]any{
		"start": map[string]any{
			"project_id": projectID,
			"op_name":    opName,
			"started_at": time.Now().UTC().Format(time.RFC3339Nano),
			"attributes": baseAttributes,
			"inputs":     inputs,
		},
	})
	return started["id"].(string)
}

func endCall(callID string, output map[string]any) {
	postJSON("/call/end", map[string]any{
		"end": map[string]any{
			"project_id": projectID,
			"id":         callID,
			"ended_at":   time.Now().UTC().Format(time.RFC3339Nano),
			"summary":    map[string]any{},
			"output":     output,
		},
	})
}

func main() {
	var missing []string
	for _, k := range []string{"WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"} {
		if os.Getenv(k) == "" {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		fatal("Missing required env vars: %s. See ../README.md#setup.", strings.Join(missing, ", "))
	}

	// Create three Calls — same shape as recipe 01, just repeated.
	questions := []struct{ Question, Answer string }{
		{"What is the capital of France?", "Paris"},
		{"What is the capital of Spain?", "Madrid"},
		{"What is the capital of Italy?", "Rome"},
	}
	type call struct{ id, ref, answer string }
	var calls []call
	for i, q := range questions {
		callID := startCall(fmt.Sprintf("recipe-06-call-%d", i+1), map[string]any{"question": q.Question})
		endCall(callID, map[string]any{"answer": q.Answer})
		calls = append(calls, call{callID, fmt.Sprintf("weave:///%s/call/%s", projectID, callID), q.Answer})
		fmt.Printf("Call %d: id=%s\n", i+1, callID)
	}

	// Build the batch — note + scorer feedback per Call (6 items total).
	var batch []map[string]any
	for _, c := range calls {
		batch = append(batch,
			map[string]any{
				"project_id":    projectID,
				"weave_ref":     c.ref,
				"feedback_type": noteType,
				"payload":       map[string]any{"note": fmt.Sprintf("Reviewed — answer: '%s'", c.answer)},
			},
			map[string]any{
				"project_id":    projectID,
				"weave_ref":     c.ref,
				"feedback_type": scorerType,
				"payload":       map[string]any{"output": map[string]any{"score": 1.0, "reason": fmt.Sprintf("Answer '%s' matches expected", c.answer)}},
			},
		)
	}

	// Single round trip for all six items.
	resp := postJSON("/feedback/batch/create", map[string]any{"batch": batch})
	results, _ := resp["res"].([]any)
	if len(results) != len(batch) {
		fatal("batch size mismatch: sent %d got %d", len(batch), len(results))
	}
	for i, item := range batch {
		res, _ := results[i].(map[string]any)
		fmt.Printf("Batch->Feedback: type=%s feedback_id=%v\n", item["feedback_type"], res["id"])
	}

	// --- verification ---
	// For each Call, query feedback by weave_ref and assert both the note and
	// the scorer feedback landed with the expected payload. Brief retry
	// tolerates eventual consistency in the read path.
	expectedTypes := []string{noteType, scorerType}
	for _, c := range calls {
		expectedNote := map[string]any{"note": fmt.Sprintf("Reviewed — answer: '%s'", c.answer)}
		expectedScorer := map[string]any{"output": map[string]any{"score": 1.0, "reason": fmt.Sprintf("Answer '%s' matches expected", c.answer)}}
		byType := map[string]map[string]any{}
		for i := 0; i < 5; i++ {
			res := postJSON("/feedback/query", map[string]any{
				"project_id": projectID,
				"query": map[string]any{
					"$expr": map[string]any{
						"$eq": []any{
							map[string]any{"$getField": "weave_ref"},
							map[string]any{"$literal": c.ref},
						},
					},
				},
			})
			byType = map[string]map[string]any{}
			if result, ok := res["result"].([]any); ok {
				for _, row := range result {
					if r, ok := row.(map[string]any); ok {
						if t, ok := r["feedback_type"].(string); ok {
							byType[t] = r
						}
					}
				}
			}
			have := true
			for _, t := range expectedTypes {
				if _, ok := byType[t]; !ok {
					have = false
					break
				}
			}
			if have {
				break
			}
			time.Sleep(time.Second)
		}
		for _, t := range expectedTypes {
			if _, ok := byType[t]; !ok {
				fatal("FAIL: feedback for %s not all visible after 5 reads", c.ref)
			}
		}
		if !reflect.DeepEqual(byType[noteType]["payload"], expectedNote) {
			fatal("note payload for %s: %v", c.id, byType[noteType]["payload"])
		}
		if !reflect.DeepEqual(byType[scorerType]["payload"], expectedScorer) {
			fatal("scorer payload for %s: %v", c.id, byType[scorerType]["payload"])
		}
		for _, row := range byType {
			if row["weave_ref"] != c.ref {
				fatal("weave_ref drift: %v", row["weave_ref"])
			}
		}
	}
	fmt.Printf("Verified: %d batched feedback items across %d Calls (note + scorer each)\n", len(batch), len(calls))
}
